src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -0,0 +1,402 @@
+            }
+
+            // See if there was a group provided that was exclusive and had items in it.  If so, then
+            // that's all we'll return.
+            var firstExclusiveList = completionLists.FirstOrDefault(t => t.IsExclusive && t.Items.Any());
+
+            if (firstExclusiveList != null)
+            {
+                return MergeAndPruneCompletionLists(SpecializedCollections.SingletonEnumerable(firstExclusiveList), defaultItemSpan);
+            }
+
+            // If no exclusive providers provided anything, then go through the remaining
+            // triggered list and see if any provide items.
+            var nonExclusiveLists = completionLists.Where(t => !t.IsExclusive).ToList();
+
+            // If we still don't have any items, then we're definitely done.
+            if (!nonExclusiveLists.Any(g => g.Items.Any()))
+            {
+                return null;
+            }
+
+            // If we do have items, then ask all the other (non exclusive providers) if they
+            // want to augment the items.
+            var usedProviders = nonExclusiveLists.Select(g => g.Provider);
+            var nonUsedProviders = providers.Except(usedProviders);
+            var nonUsedNonExclusiveLists = new List<CompletionContext>();
+            foreach (var provider in nonUsedProviders)
+            {
+                var completionList = await GetProviderCompletionsAsync(provider, document, caretPosition, defaultItemSpan, trigger, options, cancellationToken).ConfigureAwait(false);
+                if (completionList != null && !completionList.IsExclusive)
+                {
+                    nonUsedNonExclusiveLists.Add(completionList);
+                }
+            }
+
+            var allProvidersAndLists = nonExclusiveLists.Concat(nonUsedNonExclusiveLists).ToList();
+            if (allProvidersAndLists.Count == 0)
+            {
+                return null;
+            }
+
+            // Providers are ordered, but we processed them in our own order.  Ensure that the
+            // groups are properly ordered based on the original providers.
+            allProvidersAndLists.Sort((p1, p2) => completionProviderToIndex[p1.Provider] - completionProviderToIndex[p2.Provider]);
+
+            return MergeAndPruneCompletionLists(allProvidersAndLists, defaultItemSpan);
+        }
+
+        private CompletionList MergeAndPruneCompletionLists(IEnumerable<CompletionContext> completionLists, TextSpan contextSpan)
+        {
+            var displayNameToItemsMap = new Dictionary<string, List<CompletionItem>>();
+            CompletionItem suggestionModeItem = null;
+
+            foreach (var completionList in completionLists)
+            {
+                Contract.Assert(completionList != null);
+
+                foreach (var item in completionList.Items)
+                {
+                    Contract.Assert(item != null);
+
+                    // New items that match an existing item will replace it.
```

Comment:
Comment is now only sort of true.  In general/default the latter one is the better one.  There are only a few VB exceptions to this.